packages/flutter_tools/lib/src/custom_devices/custom_device.dart

```diff
@@ -0,0 +1,724 @@
+  }
+
+  @override
+  String get name => _config.label;
+
+  @override
+  DevicePortForwarder get portForwarder {
+    if (_config.usesPortForwarding) {
+      return CustomDevicePortForwarder(
+        deviceName: name,
+        forwardPortCommand: _config.forwardPortCommand,
+        forwardPortSuccessRegex: _config.forwardPortSuccessRegex,
+        processManager: _processManager,
+        logger: _logger,
+      );
+    }
+
+    return const NoOpDevicePortForwarder();
+  }
+
+  @override
+  Future<String> get sdkNameAndVersion => Future<String>.value(_config.sdkNameAndVersion);
+
+  @override
+  Future<LaunchResult> startApp(
+    covariant ApplicationPackage package, {
+    String mainPath,
+    String route,
+    DebuggingOptions debuggingOptions,
+    Map<String, dynamic> platformArgs,
+    bool prebuiltApplication = false,
+    bool ipv6 = false,
+    String userIdentifier
+  }) async {
+    final String assetBundleDir = getAssetBuildDirectory();
+
+    // build the asset bundle
+    await BundleBuilder().build(
+      platform: await targetPlatform,
+      buildInfo: debuggingOptions.buildInfo,
+      mainPath: mainPath,
+      depfilePath: defaultDepfilePath,
+      assetDirPath: assetBundleDir,
+      treeShakeIcons: false,
+    );
+
+    // if we have a post build step (needed for some embedders), execute it
+    if (_config.postBuildCommand != null) {
+      await _tryPostBuild(
+        appName: package.name,
+        localPath: assetBundleDir,
+      );
+    }
+
+    // install the app on the device
+    // (will invoke the uninstall and then the install command internally)
+    await installApp(package, userIdentifier: userIdentifier);
+
+    // finally launch the app
+    return _getOrCreateAppSession(package).start(
+      mainPath: mainPath,
+      route: route,
+      debuggingOptions: debuggingOptions,
+      platformArgs: platformArgs,
+      prebuiltApplication: prebuiltApplication,
+      ipv6: ipv6,
+      userIdentifier: userIdentifier,
+    );
+  }
+
+  @override
+  Future<bool> stopApp(covariant ApplicationPackage app, {String userIdentifier}) {
+    return _getOrCreateAppSession(app).stop();
+  }
+
+  @override
+  // TODO(ardera): Allow configuring or auto-detecting the target platform, https://github.com/flutter/flutter/issues/78151
+  Future<TargetPlatform> get targetPlatform async => TargetPlatform.linux_arm64;
+
+  @override
+  Future<bool> uninstallApp(covariant ApplicationPackage app, {String userIdentifier}) {
+    return _tryUninstall(appName: app.name);
+  }
+}
+
+class CustomDevices extends PollingDeviceDiscovery {
+  CustomDevices({
+    @required FeatureFlags featureFlags,
+    @required ProcessManager processManager,
+    @required Logger logger,
+  }) : _customDeviceWorkflow = CustomDeviceWorkflow(
+        featureFlags: featureFlags,
+      ),
+       _logger = logger,
+       _processManager = processManager,
+       _config = null,
+       super('custom devices');
+
+  CustomDevices.test({
```

Comment:
If you can, mark this as `@visibleForTesting` and document any defaults

Reply:
I marked it `visibleForTesting` and documented it, did you mean that by documenting any defaults?

Reply:
yeah, sg